crypto/digestible/README.md

```diff
@@ -1,175 +1,484 @@
+    fn append_to_transcript(&self, context: &[u8], transcript: &mut Transcript) {
+        match self {
+            ...
+            Self::%variant_possibility_name(val) => {
+                transcript.append_bytes(context, "var");
+                transcript.append_bytes("name", "%variant_name");
+                (self.which() as u64).append_to_transcript("which", transcript);
+                val.append_to_transcript("%variant_possibility_name", transcript);
+            },
+            ...
+        }
+    }
+```
+
+In rust, an `enum` may have no associated data. (In documentation they call this a unit variant).
+In this case, `val.append_to_transcript("%variant_possibility_name", transcript)` should simplify to
+```
+transcript.append_bytes("%variant_possbility_name", "");
+```
+
+In rust, a definition of an `enum` may implicitly declare anonymous structs and tuples associated
+to an enumerator. In this case, we follow the rules for an aggregate when appending
+the anonymous struct to the transcript, and use the empty string for its name.
+
+In rust, an enum where every variant has no associated data can be tagged with e.g. `repr(u32)`
+and interpretted directly as a `u32`, as in C enums. An implementor may reasonably decide
+to implement `digestible` for such an enum by converting to a `u32` and then appending that
+as a primitive. In this case `derive(Digestible)` should not be used, and this choice should be documented
+to allow cross-language implementations to do the same.
+
+Examples:
+...
+
+
+
+Schema Evolution:
+=================
+
+One of the main goals of this scheme is to support *schema evolution*, which means
+that just as with protobuf, we can add fields to our structures without breaking
+compatibillity. In this case, this means *ledger-compatibility* -- we would like to
+be able to add new fields to e.g. the `TxOut` structure or the `BlockContents` structure
+without changing the hashes of transactions or old blocks that don't have the new fields.
+
+The main ideas that we have to support this are:
+
+- Rust `core::option::Option` type is treated specially -- it *does not* get mapped
+to a `var` AST node, as rust enums with `derive(Digestible)` do. Instead, when
+an `Option` is visited, we append nothing if the value is `None`, and simply append the
+value when the option is `Some`. So it is "transparent" from the point of view of the
+digestible AST.
+
+- The `agg` AST node *does not* include the number of fields as part of the digest.
+  Instead, as many fields as needed are appended, and then there is a `closer` that
+  is appended to the transcript.
+
+Together this means that:
+- New `Option` fields may be added to existing structures without breaking ledger compatibility.
+- Old fields that were not `Option` may be made optional without breaking ledger compatibility.
+
+Note that fields may not be re-ordered or renamed.
+
+However, in the future, we may support a proc-macro attribute in digestible-derive that allows
```

Comment:
Cool, that would be useful.

Tangentially, the Jackson JSON processor in Java might be a good source of inspiration. That supports a variety of ways to annotate a class in order to customize how it gets serialized, e.g. https://www.baeldung.com/jackson-annotations